File: src/main/java/me/viniciusarnhold/altaria/command/MessageUtils.java
package me.viniciusarnhold.altaria.command;

import com.diffplug.common.base.Errors;
import me.viniciusarnhold.altaria.command.interfaces.ICommand;
import me.viniciusarnhold.altaria.core.BotManager;
import me.viniciusarnhold.altaria.utils.Logs;
import org.apache.commons.lang3.StringUtils;
import org.apache.logging.log4j.Logger;
import org.jetbrains.annotations.NotNull;
import sx.blah.discord.handle.obj.IChannel;
import sx.blah.discord.handle.obj.IDiscordObject;
import sx.blah.discord.handle.obj.IMessage;
import sx.blah.discord.handle.obj.IUser;
import sx.blah.discord.util.*;

import java.time.LocalDateTime;
import java.util.function.Supplier;

import static me.viniciusarnhold.altaria.utils.Timers.messageDeletionService;

/**
 * Created by Vinicius.
 *
 * @since ${PROJECT_VERSION}
 */
public class MessageUtils {

    private static final String ICON_URL = "https://vignette2.wikia.nocookie.net/pokemon--shuffle/images/e/e2/334.png/revision/latest?cb=20150912015748";

    private MessageUtils() {

    }


    public static boolean isMyCommand(@NotNull IMessage message, @NotNull ICommand thiz) {
        return isMyCommand(message.getContent(), thiz);
    }

    public static boolean isMyCommand(String message, @NotNull ICommand thiz) {
        message = message.trim();
        @NotNull final String prefix = Prefixes.getInstance().current();
        if (!message.startsWith(prefix)) {
            return false;
        }
        if (StringUtils.startsWithIgnoreCase(message, prefix + thiz.command())) {
            return true;
        }

        for (String alias : thiz.aliases()) {
            if (StringUtils.startsWithIgnoreCase(message, prefix + alias)) {
                return true;
            }
        }
        return false;
    }

    public static EmbedBuilder getEmbedBuilder() {
        return new EmbedBuilder()
                .withAuthorName(BotManager.BOT_NAME)
                .withAuthorUrl(BotManager.REPO_URL)
                .withColor(36, 177, 222)
                .withAuthorIcon(ICON_URL)
                .withTimestamp(LocalDateTime.now());
    }

    public static EmbedBuilder getEmbedBuilder(@NotNull IUser user) {
        return new EmbedBuilder()
                .withAuthorName(BotManager.BOT_NAME)
                .withAuthorUrl(BotManager.REPO_URL)
                .withColor(36, 177, 222)
                .withAuthorIcon(ICON_URL)
                .withFooterIcon(user.getAvatarURL())
                .withFooterText("Requested by @" + user.getName())
                .withTimestamp(LocalDateTime.now());
    }

    public static RequestBuilder getDefaultRequestBuilder(@NotNull IDiscordObject obj) {
        return new RequestBuilder(obj.getClient())
                .shouldBufferRequests(true)
                .onTimeout(() -> Logs.forMessage("Timeout while executing requests on default requestBuilder"))
                .setAsync(true)
                .shouldFailOnException(false);
    }

    public static MessageBuilder getMessageBuilder(@NotNull IMessage message) {
        return getMessageBuilder(message.getChannel());
    }

    public static MessageBuilder getMessageBuilder(@NotNull IChannel channel) {
        return new MessageBuilder(channel.getClient())
                .withChannel(channel);
    }

    public static MessageBuilder getSimpleMentionMessage(@NotNull IMessage message) {
        return getMessageBuilder(message)
                .withContent(message.getAuthor().mention())
                .appendContent(" ");
    }

    @NotNull
    public static <T> T retry(@NotNull Supplier<T> supplier) {
        @org.jetbrains.annotations.Nullable T value = null;
        do {
            try {
                value = supplier.get();
            } catch (Exception rle) {
                if (!(rle instanceof RateLimitException)) {
                    BotManager.LOGGER.error(rle);
                    throw new RuntimeException(rle);
                }
                try {
                    Thread.sleep(((RateLimitException) rle).getRetryDelay());
                } catch (InterruptedException e) {
                    BotManager.LOGGER.error(e);
                }
            }
        } while (value == null);
        return value;
    }

    public static void handleDiscord4JException(@NotNull Logger logger, Exception e) {
        logger.error(e);
    }

    public static <T> T supress(@NotNull Supplier<T> supplier) {
        try {
            return supplier.get();
        } catch (Exception e) {
            BotManager.LOGGER.error(e);
            throw Errors.asRuntime(e);
        }
    }

    public static void handleDiscord4JException(@NotNull Logger logger, @NotNull Exception e, @NotNull ICommand commandHandler, @NotNull IMessage message) {
        try {
            logger.error("{} failed to handle command, a {} was captured",
                    () -> commandHandler.getClass().getSimpleName(),
                    () -> e.getClass().getSimpleName());
            logger.error(e);

            if (e instanceof MissingPermissionsException) {
                messageDeletionService().schedule(getMessageBuilder(message)
                        .appendContent(message.getAuthor().mention())
                        .appendContent(" I dont have the necessary permissions to execute that action,")
                        .appendContent(" please give me the following permissions and try again")
                        .appendContent(System.lineSeparator())
                        .appendContent(((MissingPermissionsException) e).getMissingPermissions().toString())
                        .send());
            }

        } catch (@NotNull RateLimitException | DiscordException | MissingPermissionsException e1) {
            logger.error(e1);
        }
    }
}

File: commands/src/main/kotlin/commands/MessageUtils.kt
package me.viniciusarnhold.altaria.command

import me.viniciusarnhold.altaria.core.BotManager
import me.viniciusarnhold.altaria.utils.Logs
import me.viniciusarnhold.altaria.utils.Timers
import org.apache.commons.lang3.StringUtils
import org.apache.logging.log4j.Logger
import sx.blah.discord.handle.obj.IChannel
import sx.blah.discord.handle.obj.IDiscordObject
import sx.blah.discord.handle.obj.IMessage
import sx.blah.discord.handle.obj.IUser
import sx.blah.discord.util.*
import java.awt.Color
import java.time.LocalDateTime
import java.util.function.Supplier

/**
 * Created by Vinicius.

 * @since ${PROJECT_VERSION}
 */
object MessageUtils {

    private val ICON_URL = "https://vignette2.wikia.nocookie.net/pokemon--shuffle/images/e/e2/334.png/revision/latest?cb=20150912015748"


    fun isMyCommand(message: IMessage, thiz: IMessageCommand): Boolean {
        return isMyCommand(message.content, thiz)
    }

    fun isMyCommand(message: String, thiz: IMessageCommand): Boolean {
        var message = message
        message = message.trim { it <= ' ' }
        val prefix = Prefixes.current()
        if (!message.startsWith(prefix)) {
            return false
        }
        if (StringUtils.startsWithIgnoreCase(message, prefix + thiz.command())) {
            return true
        }

        for (alias in thiz.aliases()) {
            if (StringUtils.startsWithIgnoreCase(message, prefix + alias)) {
                return true
            }
        }
        return false
    }

    val embedBuilder: EmbedBuilder
        get() = EmbedBuilder()
                .withAuthorName(BotManager.BOT_NAME)
                .withAuthorUrl(BotManager.REPO_URL)
                .withColor(Color(36, 177, 222))
                .withAuthorIcon(ICON_URL)
                .withTimestamp(LocalDateTime.now())

    fun getEmbedBuilder(user: IUser): EmbedBuilder {
        return EmbedBuilder()
                .withAuthorName(BotManager.BOT_NAME)
                .withAuthorUrl(BotManager.REPO_URL)
                .withColor(Color(36, 177, 222))
                .withAuthorIcon(ICON_URL)
                .withFooterIcon(user.avatarURL)
                .withFooterText("Requested by @" + user.name)
                .withTimestamp(LocalDateTime.now())
    }

    fun getDefaultRequestBuilder(obj: IDiscordObject<*>): RequestBuilder {
        return RequestBuilder(obj.getClient())
                .shouldBufferRequests(true)
                .onTimeout { Logs.forMessage("Timeout while executing requests on default requestBuilder") }
                .setAsync(true)
                .shouldFailOnException(false)
    }

    fun getMessageBuilder(message: IMessage): MessageBuilder {
        return getMessageBuilder(message.channel)
    }

    fun getMessageBuilder(channel: IChannel): MessageBuilder {
        return MessageBuilder(channel.client)
                .withChannel(channel)
    }

    fun getSimpleMentionMessage(message: IMessage): MessageBuilder {
        return getMessageBuilder(message)
                .withContent(message.author.mention())
                .appendContent(" ")
    }

    fun <T> retry(supplier: Supplier<T>): T {
        var value: T? = null
        do {
            try {
                value = supplier.get()
            } catch (rle: Exception) {
                if (rle !is RateLimitException) {
                    BotManager.LOGGER.error(rle)
                    throw RuntimeException(rle)
                }
                try {
                    Thread.sleep(rle.retryDelay)
                } catch (e: InterruptedException) {
                    BotManager.LOGGER.error(e)
                }
            }
        } while (value == null)
        return value
    }

    fun handleDiscord4JException(logger: Logger, e: Exception) {
        logger.error(e)
    }

    fun <T> supress(supplier: Supplier<T>): T {
        try {
            return supplier.get()
        } catch (e: Exception) {
            BotManager.LOGGER.error(e)
            throw RuntimeException(e)
        }

    }

    fun handleDiscord4JException(logger: Logger, e: Exception, commandHandler: IMessageCommand, message: IMessage) {
        try {
            logger.error("{} failed to handle command, a {} was captured",
                    { commandHandler.javaClass.simpleName },
                    { e.javaClass.simpleName })
            logger.error(e)

            if (e is MissingPermissionsException) {
                Timers.MessageDeletionService.schedule(getMessageBuilder(message)
                        .appendContent(message.author.mention())
                        .appendContent(" I dont have the necessary permissions to execute that action,")
                        .appendContent(" please give me the following permissions and try again")
                        .appendContent(System.lineSeparator())
                        .appendContent(e.missingPermissions.toString())
                        .send())
            }

        } catch (e1: RateLimitException) {
            logger.error(e1)
        } catch (e1: DiscordException) {
            logger.error(e1)
        } catch (e1: MissingPermissionsException) {
            logger.error(e1)
        }

    }
}
